### PR TITLE
Notification types

### PR DESCRIPTION
**Pull request by SylvainCorlay**
As per discussion in #48, this enables multiple notification types in traitlets.
___
First, the `change` dictionary passed to registered observers
- must contain a `'type'` key
- potentially other keys depending on the value for `type`. For example, when the value for the `'type'` key is `'trait_change'`, we must provide the usual `'owner'`, `'old'`, `'new'`, and `'name'`.

The goal is to enable other types of notifications potentially implementing events for changes of elements in containers like `List` and `Dict`. For each type, we may have a different protocol.
___
The second thing that is implemented is the ability to filter by notification type when registering an observer. See the following example

```Python
class Foo(HasTraits):
    bar = Int()
    baz = EventfulList()

    @observe('bar', type='trait_change')  # Filter to only observe trait changes
    def handler_bar(self, change):
        pass

    @observe('baz ', type='element_change')  # Register to element change notifications for baz
    def handler_baz(self, change):
        pass

    @observe('bar', 'baz')  # Register to all notification types for `bar` and `baz` 
    def handler_all(self, change):
        pass
```

Similarly, the `observe` method also takes a `type` keyword argument defaulting to `None`, which corresponds to all types of notifications.

In terms of implementation, the `_trait_notifiers` dict is now a dict of dict, the top-level key being the trait name and the second-level being the notification type.
___
`_notify_trait` was refactored into a generic `_notify_change` that is valid for any type of notification, and the `_hold_trait_notifications` context manager was updated accordingly.
___
Another minor change is that the `observe` method' s argument `name` is renamed to `names`, which is more consistent with what it does.

### CODE REVIEW

**Comment by Carreau** [comment]
Marking as wishlist for now until not-wip for quick filtering.

**Comment by SylvainCorlay** [comment]
Now comes on the top of #85.

**Comment by SylvainCorlay** [comment]
I just added a first implementation of the filtering by notification type.

**Comment by SylvainCorlay** [comment]
@jasongrout I am still wondering if the default value for `type`  should not be `all` (i.e. None) rather than `trait_change`.

**Comment by jasongrout** [comment]
I think it should probably be all.  There are no other trait notifications now, right?

**Comment by jasongrout** [comment]
What if we want to trigger on two types?  Set it to all and manually filter?  Or can we give a list of types?

**Comment by SylvainCorlay** [comment]
@jasongrout now defaulting to `all`.

**Comment by SylvainCorlay** [comment]
TODO: 
- [ ] separate `_trait_notify` into 
    - a generic, type-independent `_notify_change` that takes the `type` and the `change` dict.
    - the creation of the change dictionary that should be made in `TraitType.__set__`
- [ ] replace the `merge` defined in `hold_trait_notification` with a `compress`. compress is identity on two `change` dictionaries having different type keys. compress merges trait_change notifications (ie creates a single message to send) when both types are `trait_change` .

**Comment by minrk** [comment]
@SylvainCorlay can you summarize a bit of what the changes are, and what they will be used for in the description here, so people don't have to search through #48 to find the relevant bits?

I think test-as-demo would also be useful.

**Comment by SylvainCorlay** [comment]
@minrk I have updated the description of the issue and iterated on it.

There is still work to do. The crux of what remains is that the `_hold_trait_notifications` context manager knows how to merge `trait_change` events into a single one but not notifications of different types.

**Comment by SylvainCorlay** [comment]
So the main question is how `hold_trait_notifications` will roll back things in case of TraitError. I updated the logic so that it roll backs `trait_change`-typed notifications, but we cannot do much more so far. Removing the [WIP] tag.

**Comment by sccolbert** [comment]
Are the signatures for those handlers correct? Should they just be `handler(self, change)` instead of `handler(self, **change)`?

**Comment by SylvainCorlay** [comment]
I think it is `handler(change)` everywhere.

**Comment by minrk** [comment]
Question: is the default for `@observe` all events or just trait_change? It looks like it's all events, but I suspect that the vast majority of cases will only want to observe the change event, in which case perhaps the default should be the change event, and you would have to explicitly ask for all events. What do you think @SylvainCorlay?

Also, reading the code, the recent change to to make `None` the magic value that indicates all types makes it harder to track what's going on. Perhaps a `traitlets.AllEvents` sentinel would make that clearer.

**Comment by SylvainCorlay** [comment]
@minrk indeed the default value to `type` is `None`, which corresponds to all events. 
Same for `name`, where `None` corresponds to the late `'all_traits'` string.

I initially defaulted to `trait_change`, but @jasongrout convinced me to default to all.

I agree that in the large majority of cases will be about change events. However
- when you specify the `name`, and the targeted decorator only emits change notifications, there should not be issues...
- in the case where you don't specify the `name` but your instance only has old-style trait types, there should not be issues either.

The case where people have to be careful is when adding descriptors that emit other types and notifications where they have listeners listening to all notifications...

Regarding the new special value like `traitlets.AllEvents`, I am :+1: . Could it be the same `traitlets.All` for event types and trait names?

**Comment by minrk** [comment]
> Could it be the same traitlets.All

Seems like a good idea, though when I look at the code samples above, since `type` is a string in all other cases, it seems more consistent for `type='all'` to be the way to specify all events, rather than a singleton non-string. What do you think?

> The case where people have to be careful is when adding descriptors that emit other types and notifications where they have listeners listening to all notifications...

But isn't that going to be ~all of the time? In most cases, people are going to register `@observe` with the default events (if the default isn't the most common, the default is probably wrong). If events other than change are rare and the default includes all events, then I would expect most cases to register for all events but only handle change events, in which case emitting any other event will result in errors.

**Comment by SylvainCorlay** [comment]
> > Could it be the same traitlets.All

> Seems like a good idea, though when I look at the code samples above, since type is a string in all other cases, it seems more consistent for type='all' to be the way to specify all events, rather than a singleton non-string. What do you think?

It is also nice to use the same thing for `type` and `name`.

> > The case where people have to be careful is when adding descriptors that emit other types and notifications where they have listeners listening to all notifications...

> But isn't that going to be ~all of the time? In most cases, people are going to register @observe with the default events (if the default isn't the most common, the default is probably wrong). If events other than change are rare and the default includes all events, then I would expect most cases to register for all events but only handle change events, in which case emitting any other event will result in errors.

I think that people use `@observe` with a specific name, and rarely listen to change of all trait attributes in the same handler.

**Comment by minrk** [comment]
> I think that people use @observe with a specific name most of the time, and rarely listen to changes of all trait attributes with the same handler.

Gotcha, so the default behavior of `@observe('bar')` should be to only observe the `trait_change` event of `bar`, then?

**Comment by SylvainCorlay** [comment]
If 'bar' does not emit anything else than `trait_change` events...

**Comment by minrk** [comment]
@SylvainCorlay okay, so if I do `@observe('bar')` and only expect change events, some future change could add another event that my handler will be called with, and will result in an error. That doesn't seem like the right thing to do, but maybe I don't understand the alternate events, since none of them exist. Do you think it's going to be more common for people to want to register a single handler for all events and all their signatures, or just the change event? If it's just the change event, that should probably be the default behavior.

**Comment by SylvainCorlay** [comment]
Yes we are on the same page. A way to mitigate that issue is to make would be to make it so that the current existing trait type never emit other events (and create EventFulList type that has other events later).
I am also fine with defaulting back to `trait_change`.

**Comment by minrk** [comment]
It would also mean that adding any new event types would be a breaking change, since you have to assume that people are doing `@observe` with no arguments, and will get a new event that they can't handle, whereas with `@observe('bar', type='all')` it's clear that multiple event types may come in. If there were more different events that were likely to be common, I wouldn't be as concerned, but since change is likely to be the only event in so many cases, default=all worries me a bit.

@jasongrout are you okay with trait_change being the default? It seems more logical to me, but the two of you are more familiar with the motivation behind the change, so I might be misunderstanding.

**Comment by ellisonbg** [comment]
Yeah, I think that it probably makes more sense to default to `type=trait_change`. Also, why not call that type `change`?

**Comment by SylvainCorlay** [comment]
- defaulted back to `type='trait_change'`
- created a `All` sentinel value to replace `None` as a key corresponding to all notification types or all trait attributes.

Although I am still not sure that we should default to `trait_change`. Strictly speaking, defaulting to `All` is backward compatible as of now, but it will be impossible to make the change in a backward compatible fashion later...
